### ISSUE TITLE
bindgen fails to load libclang.dll

### ISSUE THREAD

**Issue by Boscop** [KyleMayes/clang-sys#74]
From https://github.com/rust-lang-nursery/rust-bindgen/issues/1276

Why can't bindgen find libclang.dll? 
```
D:\3rdparty\reaper_extension_sdk\jmde>bindgen reaper_plugin_functions.h -o reaper.rs
thread 'main' panicked at 'Unable to find libclang: "the `libclang` shared libra
ry at D:\\Program Files\\LLVM\\bin\\libclang.dll could not be opened: %1 is not
a valid Win32 application. (os error 193)"', libcore\result.rs:945:5
stack backtrace:
   0: <unknown>
   1: <unknown>
   2: <unknown>
   3: <unknown>
   4: <unknown>
   5: <unknown>
   6: <unknown>
   7: <unknown>
   8: <unknown>
   9: <unknown>
  10: <unknown>
  11: <unknown>
  12: <unknown>
  13: <unknown>
  14: BaseThreadInitThunk
  15: RtlInitializeExceptionChain
  16: RtlInitializeExceptionChain
```

Btw, the file "D:\Program Files\LLVM\bin\libclang.dll" exists and has version 4.0.0.0.
An earlier version of bindgen cli was working without problems on this system.

**Comment by ceph3us** [comment]
I believe I've managed to narrow down this issue while working on sodiumoxide/sodiumoxide/pull/200:

- The Windows search paths are as follows:
```
const SEARCH_WINDOWS: &[&str] = &[
    "C:\\LLVM\\lib",
    "C:\\Program Files*\\LLVM\\lib",
    "C:\\MSYS*\\MinGW*\\lib",
];
```
- The search assumes the first DLL it finds is the correct one.
- If you are compiling on x64 Windows, and have both x64 and x86 Rust MSVC toolchains installed, it will always choose `C:\Program Files\LLVM\bin\libclang.dll`, which is the 64-bit DLL on such machines.
- If cross compiling for `i686-pc-windows-msvc` on x64 Windows, the 32-bit Rust compiler will attempt to link against the 64-bit `libclang.dll`, which will obviously fail.

**Comment by ceph3us** [comment]
Not the exact same message, but the same cause and symptom:
https://ci.appveyor.com/project/ceph3us/clang-sys/build/1.0.3

**Comment by KyleMayes** [comment]
Thanks for the investigation and root cause!

The build script already filters out binaries for the wrong architecture on Linux/FreeBSD ([here](https://github.com/KyleMayes/clang-sys/blob/e6948cb8e7d6d67d6eb409aa788699b401ef9585/build.rs#L138-L162)). This will need to be extended to also filter out binaries for the wrong architecture on Windows.

It looks like from my cursory investigation that this information is in the PE header for the binary and that it shouldn't be too difficult to add this functionality.

I'll try to get this done over the weekend.

**Comment by KyleMayes** [comment]
Could you try out my branch of `rust-bindgen` [here](https://github.com/KyleMayes/rust-bindgen/tree/win32-64) with your project and see if that solves the problem? This branch has a (hopefully) fixed version of `clang-sys` as a dependency.

Thanks.

**Comment by KyleMayes** [comment]
I have released `v0.23.0` with the architecture detection for DLLs.

If this does not solve your problem, please reopen this issue or open another.
